go/service/kvstore.go

```diff
@@ -0,0 +1,239 @@
+	ctx = libkb.WithLogTag(ctx, "KV")
+	mctx := libkb.NewMetaContext(ctx, h.G())
+	defer mctx.TraceTimed(fmt.Sprintf("KVStoreHandler#PutKVEntry: t:%s, n:%s, k:%s", arg.TeamName, arg.Namespace, arg.EntryKey), func() error { return err })()
+
+	if err := h.assertLoggedIn(ctx); err != nil {
+		mctx.Debug("not logged in err: %v", err)
+		return res, err
+	}
+	teamID, err := h.resolveTeam(mctx, arg.TeamName)
+	if err != nil {
+		mctx.Debug("error resolving team with name %s: %v", arg.TeamName, err)
+		return res, err
+	}
+	entryID := keybase1.KVEntryID{
+		TeamID:    teamID,
+		Namespace: arg.Namespace,
+		EntryKey:  arg.EntryKey,
+	}
+	apiRes, err := h.serverFetch(mctx, entryID)
+	if err != nil {
+		mctx.Debug("error fetching %+v from server: %v", entryID, err)
+		return res, err
+	}
+	var cleartext string
+	if apiRes.Ciphertext != "" {
+		cleartext, err = h.Boxer.Unbox(mctx, entryID, apiRes.Revision, apiRes.Ciphertext, apiRes.FormatVersion, apiRes.WriterUID, apiRes.WriterEldestSeqno, apiRes.WriterDeviceID)
+		if err != nil {
+			mctx.Debug("error unboxing %+v: %v", entryID, err)
+			return res, err
+		}
+	}
+	res = keybase1.KVGetResult{
+		TeamName:   arg.TeamName,
+		Namespace:  arg.Namespace,
+		EntryKey:   arg.EntryKey,
+		EntryValue: cleartext,
+		Revision:   apiRes.Revision,
+	}
+	return res, nil
+}
+
+type putEntryAPIRes struct {
+	Status   libkb.AppStatus `json:"status"`
+	Revision int             `json:"revision"`
+}
+
+func (k *putEntryAPIRes) GetAppStatus() *libkb.AppStatus {
+	return &k.Status
+}
+
+func (h *KVStoreHandler) fetchRevisionFromCacheOrServer(mctx libkb.MetaContext, entryID keybase1.KVEntryID) (int, error) {
+	_, _, prevRevision := mctx.G().GetKVRevisionCache().Fetch(mctx, entryID)
+	if prevRevision == 0 {
+		// not in the cache. check if it's in the server.
+		serverRes, err := h.serverFetch(mctx, entryID)
+		if err != nil {
+			return 0, err
+		}
+		prevRevision = serverRes.Revision
+	}
+	return prevRevision, nil
+}
+
+func (h *KVStoreHandler) PutKVEntry(ctx context.Context, arg keybase1.PutKVEntryArg) (res keybase1.KVPutResult, err error) {
+	ctx = libkb.WithLogTag(ctx, "KV")
+	mctx := libkb.NewMetaContext(ctx, h.G())
+	defer mctx.TraceTimed(fmt.Sprintf("KVStoreHandler#PutKVEntry: t:%s, n:%s, k:%s", arg.TeamName, arg.Namespace, arg.EntryKey), func() error { return err })()
+	if err := h.assertLoggedIn(ctx); err != nil {
+		mctx.Debug("not logged in err: %v", err)
+		return res, err
+	}
+	teamID, err := h.resolveTeam(mctx, arg.TeamName)
+	if err != nil {
+		mctx.Debug("error resolving team with name %s: %v", arg.TeamName, err)
+		return res, err
+	}
+	entryID := keybase1.KVEntryID{
+		TeamID:    teamID,
+		Namespace: arg.Namespace,
+		EntryKey:  arg.EntryKey,
+	}
+	prevRevision, err := h.fetchRevisionFromCacheOrServer(mctx, entryID)
+	if err != nil {
+		mctx.Debug("error fetching the revision for %+v from the local cache: %v", entryID, err)
+		return res, err
+	}
+	revision := prevRevision + 1
```

Comment:
correct. i had initially planned on adding revision and conflict management to this PR, but i realized it wasn't strictly necessary for a very first pass, and it would be fairly easy to review in a single future PR. but yes. this is the idea. 
device `a` will pass in the revision with the second write and get a very helpful error message. and we can decide what to do in the case you've described; my instinct is overwrite if no revision is passed, and error with something very helpful (i.e. this entry changed since you've last seen it. you should fetch it first and take a look) if you used an old one. that way two devices can stay in sync either by both using revisions or both not using them.

Reply:
ticket so this doesn't get missed: PICNIC-534